src/menhir_parser.mly

```diff
@@ -4,16 +4,107 @@ open Lang
   %}
 
 %token EOF
+%token COLUM
+%token SEMICOL
+%token EQUAL
+%token DASH
+%token<string> STR
+%token<string> URL
+%token<int> INT
 
 %start <Lang.identifier> identifier
 
 %%
   let identifier_core :=
-      | { (1, Snapshot, [||]) }
+      | swh = STR; COLUM; scheme_v = INT; COLUM; obj_t = STR; COLUM; hash = STR; {
 
-let qualifiers :=
-    | { [] }
+    if swh <> "swh" then
+      failwith "Prefix incorrect"
 
-let identifier :=
-    | ~ = identifier_core; ~ = qualifiers; EOF; { identifier_core, qualifiers }
-                                              | ~ = identifier_core; _blah = identifier_core; ~ = qualifiers; EOF; { identifier_core, qualifiers }
+    else (
+      let obj_id = String.lowercase_ascii hash in
+
+      String.iter (fun c ->
+        match c with
+        | 'a'..'f' | '0'..'9' -> ()
+        | _ -> failwith "Object_ID is invalid"
+      ) hash;
+
+      if String.length obj_id = 40 then
+        let obj_id = Array.init 40 (String.get obj_id) in
```

Comment:
create a function `object_id_of_string` in `Lang` of type `string -> (char Array) option` and use it there so you have only two cases